README.md
What is this?
=============

I Know My ABCs is an activity for introducing the Spanish alphabet. It displays letters and images and associated sound files, such as ‘A as in ave’.

How to use?
===========

I Know My ABCs is not part of the Sugar desktop, but can be added.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/),
* [Download I Know My ABCs using Browse](https://activities.sugarlabs.org/), search for `ABCs`, then download I Know My ABCs, and;
* [How to use I Know My ABCs](https://help.sugarlabs.org/iknowmyabcs.html)

How to upgrade?
===============

On Sugar desktop systems;
* use [My Settings](https://help.sugarlabs.org/my_settings.html), [Software Update](https://help.sugarlabs.org/my_settings.html#software-update), or;
* use Browse to open [activities.sugarlabs.org](https://activities.sugarlabs.org/), search for `ABCs`, then download.


NEWS

7

* Regenerate POT file (Jui Pradhan),
* Fetch Translations (Jui Pradhan),
* Port from GObject to GLib (Jui Pradhan),
* Deanglify links to help.sugarlabs.org (James Cameron),
* Update README - not a Fructose or Honey activity (James Cameron),
* Port to Python 3 (Ayush Nawal),
* Fix source_remove warning (Ayush Nawal),

6

* Fix Audio bug
* Port setup.py to Gtk+3
* Add README
* Update Metadata

5

Enhancements based on feedback from .NI:

* Level 1: add the uppercase and lowercase letters
* Level 2: link the name of the letter and its image;
  for example: The XO says: "A" de "Ave", "E" de Elefante
* Change the background color to one of the XO colors
* Add visual to feedback for Levels 3 and 4
* Don't repeat letter until all letters seen

3

* GTK3 port by Ignacio Rodriguez

1

* New 'I Know My ABCs' activity




genpieces.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#Copyright (c) 2009-11 Walter Bender

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.

import os


class SVG:
    ''' SVG generators '''

    def __init__(self):
        self._scale = 1
        self._stroke_width = 1
        self._fill = '#FFFFFF'
        self._stroke = '#000000'

    def _svg_style(self, stroke=False, extras=""):
        if stroke:
            return "%s%s%s%s%s%f%s%s%s" % ("style=\"fill:", self._fill,
                                           ";stroke:", "#000000",
                                           ";stroke-width:",
                                           self._stroke_width, ";", extras,
                                           "\" />\n")
        else:
            return "%s%s%s%s%s%f%s%s%s" % ("style=\"fill:", self._fill,
                                           ";stroke:", self._stroke,
                                           ";stroke-width:",
                                           self._stroke_width, ";", extras,
                                           "\" />\n")

    def _svg_text(self, x, y, font_size, text_string, stroke=False,
                  center=False):
        if center:
            align = 'text-align:center;text-anchor:middle'
        else:
            align = ''
        svg_string = "<text x=\"%d\" y=\"%d\" " % (x, y)
        if stroke:
            svg_string += "style=\"font-size:%dpx;font-weight:bold;font-family:Sans;fill:%s;\
stroke:#000000;%s\">" % (font_size, self._stroke, align)
        else:
            svg_string += "style=\"font-size:%dpx;font-weight:bold;font-family:Sans;fill:%s;\
%s\">" % (font_size, self._stroke, align)
        svg_string += "<tspan x=\"%d\" y=\"%d\">%s</tspan></text>" % \
            (x, y, text_string)
        return svg_string

    def _svg_line(self, x1, y1, x2, y2):
        svg_string = "<line x1=\"%f\" y1=\"%f\" x2=\"%f\" y2=\"%f\"\n" % \
                      (x1, y1, x2, y2)
        svg_string += self._svg_style("stroke-linecap:square;")
        return svg_string

    def _svg_rect(self, w, h, rx, ry, x, y, stroke=False):
        svg_string = "       <rect\n"
        svg_string += "          width=\"%f\"\n" % (w)
        svg_string += "          height=\"%f\"\n" % (h)
        svg_string += "          rx=\"%f\"\n" % (rx)
        svg_string += "          ry=\"%f\"\n" % (ry)
        svg_string += "          x=\"%f\"\n" % (x)
        svg_string += "          y=\"%f\"\n" % (y)
        self.set_stroke_width(1.0)
        svg_string += self._svg_style(stroke=stroke)
        return svg_string

    def _background(self, scale, stroke=False):
        return self._svg_rect(79.5 * scale, 59.5 * scale, 1, 1, 0.25, 0.25,
                              stroke=stroke)

    def header(self, scale=1, background=True, stroke=False):
        svg_string = "<?xml version=\"1.0\" encoding=\"UTF-8\""
        svg_string += " standalone=\"no\"?>\n"
        svg_string += "<!-- Created with Emacs -->\n"
        svg_string += "<svg\n"
        svg_string += "   xmlns:svg=\"http://www.w3.org/2000/svg\"\n"
        svg_string += "   xmlns=\"http://www.w3.org/2000/svg\"\n"
        svg_string += "   version=\"1.0\"\n"
        svg_string += "%s%f%s" % ("   width=\"", scale * 80 * self._scale,
                                  "\"\n")
        svg_string += "%s%f%s" % ("   height=\"", scale * 60 * self._scale,
                                  "\">\n")
        svg_string += "%s%f%s%f%s" % ("<g\n       transform=\"matrix(",
                                      self._scale, ",0,0,", self._scale,
                                      ",0,0)\">\n")
        if background:
            svg_string += self._background(scale, stroke=stroke)
        return svg_string

    def footer(self):
        svg_string = "</g>\n"
        svg_string += "</svg>\n"
        return svg_string

    def set_scale(self, scale=1.0):
        self._scale = scale

    def set_colors(self, colors):
        self._stroke = colors[0]
        self._fill = colors[1]

    def set_stroke_width(self, stroke_width=1.0):
        self._stroke_width = stroke_width


def generate_card(string='a', colors=['#FF0000', '#FFFFFF'],
                  background=True, scale=1, stroke=False, center=False,
                  font_size=40):
    if len(string) > 2:
        font_size *= 0.75
    svg = SVG()
    svg.set_scale(scale)
    svg.set_colors(colors)
    svg_string = svg.header(background=background, stroke=stroke)
    if center:
        x = 40
    else:
        x = 5
    svg_string += svg._svg_text(x, 45, font_size, string, stroke=stroke,
                                    center=center)
    svg_string += svg.footer()
    return svg_string


def open_file(datapath, filename):
    return file(os.path.join(datapath, filename), "w")


def close_file(f):
    f.close()


def generator(datapath):
    i = 1
    filename = "tile-%d.svg" % (i)
    f = open_file(datapath, filename)
    f.write(generate_card(string='b', background=False))
    close_file(f)


def genblank(w, h, colors, stroke_width=1.0):
    svg = SVG()
    svg.set_scale(1)
    svg.set_colors(colors)
    svg.set_stroke_width(stroke_width)
    svg_string = svg.header(int(w / 80), int(h / 60))
    svg_string += svg.footer()
    return svg_string


def main():
    return 0

if __name__ == "__main__":
    if not os.path.exists(os.path.join(os.path.abspath('.'), 'images')):
        os.mkdir(os.path.join(os.path.abspath('.'), 'images'))
    generator(os.path.join(os.path.abspath('.'), 'images'))
    main()


IKnowMyABCs.py
#Copyright (c) 2012,13 Walter Bender
#Copyright (c) 2012 Ignacio Rodriguez

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
from gi.repository import Gdk

from sugar3.activity import activity
from sugar3.graphics.toolbarbox import ToolbarBox, ToolbarButton
from sugar3.graphics.toolbutton import ToolButton
from sugar3.activity.widgets import ActivityToolbarButton
from sugar3.activity.widgets import StopButton

from gettext import gettext as _
import os.path

from page import Page
from utils.toolbar_utils import separator_factory, label_factory, radio_factory

import logging
_logger = logging.getLogger('iknowmyabcs-activity')


class IKnowMyABCs(activity.Activity):
    ''' Learning the alphabet.

    Level1: The alphabet appears and the user has the option to click
    on a letter to listen the name of it and the sound of it.

    Level2: The laptop says a letter and the user must click on the
    correct one. '''

    def __init__(self, handle):
        ''' Initialize the toolbars and the reading board '''
        super(IKnowMyABCs, self).__init__(handle)

        if 'LANG' in os.environ:
            language = os.environ['LANG'][0:2]
        elif 'LANGUAGE' in os.environ:
            language = os.environ['LANGUAGE'][0:2]
        else:
            language = 'es'  # default to Spanish

        # FIXME: find some reasonable default situation
        language = 'es'

        self.activity_path = activity.get_bundle_path()

        self._lessons_path = os.path.join(self.activity_path,
                                          'lessons', language)
        self._images_path = os.path.join(self.activity_path,
                                          'images', language)
        self._sounds_path = os.path.join(self.activity_path,
                                          'sounds', language)

        self._setup_toolbars()

        # Create a canvas
        canvas = Gtk.DrawingArea()
        width = Gdk.Screen.width()
        height = int(Gdk.Screen.height())
        canvas.set_size_request(width, height)
        canvas.modify_bg(Gtk.StateType.NORMAL, Gdk.color_parse("#000000"))
        canvas.show()
        self.set_canvas(canvas)

        self.mode = 'letter'

        self._page = Page(canvas, self._lessons_path,
                          self._images_path, self._sounds_path,
                          parent=self)

    def _setup_toolbars(self):
        ''' Setup the toolbars.. '''

        # no sharing
        self.max_participants = 1

        toolbox = ToolbarBox()

        # Activity toolbar
        activity_button = ActivityToolbarButton(self)

        toolbox.toolbar.insert(activity_button, 0)
        activity_button.show()

        self.set_toolbar_box(toolbox)
        toolbox.show()
        primary_toolbar = toolbox.toolbar
        button = radio_factory('letter', primary_toolbar, self._letter_cb,
                               tooltip=_('Listen to the letter names.'))
        radio_factory('picture', primary_toolbar, self._picture_cb,
                      tooltip=_('Listen to the letter names.'),
                      group=button)
        '''
        radio_factory('sort', primary_toolbar, self._sort_cb,
                      tooltip=_('sort into abc order'),
                      group=button)
        '''
        radio_factory('find1', primary_toolbar, self._find1_cb,
                      tooltip=_(
                'Find the letter that corresponds to the sound.'),
                      group=button)
        radio_factory('find2', primary_toolbar, self._find2_cb,
                      tooltip=_(
                'Find the letter that corresponds to the sound.'),
                      group=button)

        self.status = label_factory(primary_toolbar, '', width=300)

        separator_factory(primary_toolbar, True, False)

        stop_button = StopButton(self)
        stop_button.props.accelerator = '<Ctrl>q'
        toolbox.toolbar.insert(stop_button, -1)
        stop_button.show()

    def _letter_cb(self, event=None):
        ''' click on card to hear the letter name '''
        self.mode = 'letter'
        self.status.set_text(_('Click on a card to hear a sound.'))
        if hasattr(self, '_page'):
            self._page.new_page()
        return

    def _picture_cb(self, event=None):
        ''' click on card to hear the letter name '''
        self.mode = 'picture'
        self.status.set_text(_('Click on a card to hear a sound.'))
        if hasattr(self, '_page'):
            self._page.new_page(cardtype='image')
        return

    def _sort_cb(self, event=None):
        ''' sort cards into ABC order '''
        self.mode = 'sort'
        if hasattr(self, '_page'):
            self._page.new_page()
            # self._page.random_order()
        return

    def _find1_cb(self, event=None):
        ''' letter sound --> letter card '''
        self.mode = 'find by letter'
        if hasattr(self, '_page'):
            _logger.debug('find1 button pushed')
            self._page.new_page()
            self._page.new_target()
        return

    def _find2_cb(self, event=None):
        ''' word sound --> letter card '''
        self.mode = 'find by word'
        if hasattr(self, '_page'):
            _logger.debug('find2 button pushed')
            self._page.new_page()
            self._page.new_target()
        return

    def write_file(self, file_path):
        ''' Write status to the Journal '''
        if not hasattr(self, '_page'):
            return
        self.metadata['page'] = str(self._page.current_card)


def get_path(activity, subpath):
    """ Find a Rainbow-approved place for temporary files. """
    try:
        return(os.path.join(activity.get_activity_root(), subpath))
    except:
        # Early versions of Sugar didn't support get_activity_root()
        return(os.path.join(
                os.environ['HOME'], ".sugar/default", SERVICE, subpath))


utils/play_audio.py
"""
 aplay.py
 Copyright (C) 2007 Andy Wingo <wingo@pobox.com>
 Copyright (C) 2007 Red Hat, Inc.
 Copyright (C) 2008-2010 Kushal Das <kushal@fedoraproject.org>
 Copyright (C) 2010-11 Walter Bender
 Copyright (C) 2013-14 Aneesh Dogra <lionaneesh@gmail.com>
"""

# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA

import gi
gi.require_version('Gst', '1.0')
from gi.repository import GLib, Gst

import logging
import time
_logger = logging.getLogger('iknowabcs-activity')

def play_audio_from_file(file_path, queue=False):
    """ Audio media """
    if hasattr(play_audio_from_file, 'player') and \
       play_audio_from_file.player:
        if queue:
            if hasattr(play_audio_from_file, 'queue'):
                if play_audio_from_file.queue and \
                   file_path in play_audio_from_file.queue:
                    # if we already have that file in the queue
                    # we'll just update the timer.
                    if hasattr(play_audio_from_file, 'queue_timeout'):
                        time.sleep(0.01)
                        GLib.source_remove(play_audio_from_file.queue_timeout)
                        f = Gst.Format(Gst.Format.TIME)
                        duration = play_audio_from_file.player.query_duration(f)[0]
                        timeout = duration / 1000000000.
                        play_audio_from_file.queue_timeout = \
                            GLib.timeout_add(int(timeout * 1000), \
                                            play_audio_from_file, file_path)
                        return
            else:
                play_audio_from_file.queue = []

            time.sleep(0.01)
            f = Gst.Format(Gst.Format.TIME)
            duration = play_audio_from_file.player.query_duration(f)[0]
            timeout = duration / 1000000000.
            play_audio_from_file.queue_timeout = GLib.timeout_add( \
                        int(timeout * 1000), play_audio_from_file, file_path)
            play_audio_from_file.queue.append(file_path)
            return
        else:
            play_audio_from_file.player.set_state(Gst.State.NULL)

    else:
        Gst.init(None)

    play_audio_from_file.player = Gst.parse_launch (
                    'filesrc location=%s ! oggdemux ! vorbisdec ! ' \
                    'audioconvert ! alsasink' % (file_path))

    if not play_audio_from_file.player:
        _logger.warning('unable to play audio file %s' % (file_path))

    else:
        play_audio_from_file.player.set_state(Gst.State.PLAYING)


utils/toolbar_utils.py
# -*- coding: utf-8 -*-
# Copyright (c) 2011, Walter Bender

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA


from gi.repository import Gtk
from sugar3.graphics.radiotoolbutton import RadioToolButton
from sugar3.graphics.toolbutton import ToolButton
from sugar3.graphics.combobox import ComboBox
from sugar3.graphics.toolcombobox import ToolComboBox


def combo_factory(combo_array, toolbar, callback, cb_arg=None,
                  tooltip=None, default=None):
    '''Factory for making a toolbar combo box'''
    combo = ComboBox()
    if tooltip is not None and hasattr(combo, 'set_tooltip_text'):
        combo.set_tooltip_text(tooltip)
    if cb_arg is not None:
        combo.connect('changed', callback, cb_arg)
    else:
        combo.connect('changed', callback)
    for i, selection in enumerate(combo_array):
        combo.append_item(i, selection, None)
    combo.show()
    toolitem = Gtk.ToolItem()
    toolitem.add(combo)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(toolitem, -1)
    else:  # or a secondary toolbar
        toolbar.props.page.insert(toolitem, -1)
    toolitem.show()
    if default is not None:
        combo.set_active(combo_array.index(default))
    return combo


def entry_factory(default_string, toolbar, tooltip=None, max=3):
    ''' Factory for adding a text box to a toolbar '''
    entry = Gtk.Entry()
    entry.set_text(default_string)
    if tooltip is not None and hasattr(entry, 'set_tooltip_text'):
        entry.set_tooltip_text(tooltip)
    entry.set_width_chars(max)
    entry.show()
    toolitem = Gtk.ToolItem()
    toolitem.add(entry)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(toolitem, -1)
    else:  # or a secondary toolbar
        toolbar.props.page.insert(toolitem, -1)
    toolitem.show()
    return entry


def button_factory(icon_name, toolbar, callback, cb_arg=None, tooltip=None,
                   accelerator=None):
    '''Factory for making toolbar buttons'''
    button = ToolButton(icon_name)
    if tooltip is not None:
        button.set_tooltip(tooltip)
    button.props.sensitive = True
    if accelerator is not None:
        button.props.accelerator = accelerator
    if cb_arg is not None:
        button.connect('clicked', callback, cb_arg)
    else:
        button.connect('clicked', callback)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(button, -1)
    else:  # or a secondary toolbar
        toolbar.props.page.insert(button, -1)
    button.show()
    return button


def radio_factory(button_name, toolbar, callback, cb_arg=None, tooltip=None,
                  group=None):
    ''' Add a radio button to a toolbar '''
    button = RadioToolButton(group=group)
    button.set_named_icon(button_name)
    if callback is not None:
        if cb_arg is None:
            button.connect('clicked', callback)
        else:
            button.connect('clicked', callback, cb_arg)
    if hasattr(toolbar, 'insert'):  # Add button to the main toolbar...
        toolbar.insert(button, -1)
    else:  # ...or a secondary toolbar.
        toolbar.props.page.insert(button, -1)
    button.show()
    if tooltip is not None:
        button.set_tooltip(tooltip)
    return button


def label_factory(toolbar, label_text, width=None):
    ''' Factory for adding a label to a toolbar '''
    label = Gtk.Label(label_text)
    label.set_line_wrap(True)
    if width is not None:
        label.set_size_request(width, -1)  # doesn't work on XOs
    label.show()
    toolitem = Gtk.ToolItem()
    toolitem.add(label)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(toolitem, -1)
    else:  # or a secondary toolbar
        toolbar.props.page.insert(toolitem, -1)
    toolitem.show()
    return label


def separator_factory(toolbar, expand=False, visible=True):
    ''' add a separator to a toolbar '''
    separator = Gtk.SeparatorToolItem()
    separator.props.draw = visible
    separator.set_expand(expand)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(separator, -1)
    else:  # or a secondary toolbar
        toolbar.props.page.insert(separator, -1)
    separator.show()


def image_factory(image, toolbar, tooltip=None):
    ''' Add an image to the toolbar '''
    img = Gtk.Image()
    img.set_from_pixbuf(image)
    img_tool = Gtk.ToolItem()
    img_tool.add(img)
    if tooltip is not None:
        img.set_tooltip_text(tooltip)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(img_tool, -1)
    else:  # or a secondary toolbar
        toolbar.props.page.insert(img_tool, -1)
    img_tool.show()
    return img

# PLEASE FIXME #
"""
def spin_factory(default, min, max, callback, toolbar):
    spin_adj = Gtk.Adjustment(default, min, max, 1, 32, 0)
    spin = gtk.SpinButton(spin_adj, 0, 0)
    spin_id = spin.connect('value-changed', callback)
    spin.set_numeric(True)
    spin.show()
    toolitem = gtk.ToolItem()
    toolitem.add(spin)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(toolitem, -1)
    else:
        toolbar.props.page.insert(toolitem, -1)
    toolitem.show()
    return spin
"""


page.py
# -*- coding: utf-8 -*-
#Copyright (c) 2012 Walter Bender
#Copyright (c) 2012 Ignacio Rodríguez

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.

from gi.repository import Gtk, Gdk, GLib, GdkPixbuf
import os
from random import uniform

from gettext import gettext as _

from utils.play_audio import play_audio_from_file

import logging
_logger = logging.getLogger('iknowmyabcs-activity')

from sugar3 import profile
from sugar3.graphics import style
GRID_CELL_SIZE = style.GRID_CELL_SIZE

from genpieces import generate_card, genblank
from utils.sprites import Sprites, Sprite


XDIM = 6
YDIM = 5
GUTTER = 4


class Page():
    ''' Pages from Infuse Reading method '''

    def __init__(self, canvas, lessons_path, images_path, sounds_path,
                 parent=None):
        ''' The general stuff we need to track '''
        self._activity = parent
        self._lessons_path = lessons_path
        self._images_path = images_path
        self._sounds_path = sounds_path

        self._colors = profile.get_color().to_string().split(',')

        self._card_data = []
        self._color_data = []
        self._image_data = []
        self._media_data = []  # (image sound, letter sound)
        self._word_data = []
        self._deja_vu = []

        # Starting from command line
        if self._activity is None:
            self._sugar = False
            self._canvas = canvas
        else:
            self._sugar = True
            self._canvas = canvas
            self._activity.show_all()

        self._canvas.add_events(Gdk.EventMask.BUTTON_PRESS_MASK)
        self._canvas.add_events(Gdk.EventMask.BUTTON_RELEASE_MASK)
        self._canvas.connect("draw", self.__draw_cb)
        self._canvas.connect("button-press-event", self._button_press_cb)
        self._canvas.connect("button-release-event", self._button_release_cb)
        self._canvas.connect("key_press_event", self._keypress_cb)
        self._width = Gdk.Screen.width()
        self._height = Gdk.Screen.height()
        self._card_width = int((self._width / XDIM)) - GUTTER * 2
        self._card_height = int((self._height - GRID_CELL_SIZE) / YDIM) \
            - GUTTER * 2
        self._grid_x_offset = int(
            (self._width - XDIM * (self._card_width + GUTTER * 2)) / 2)
        self._grid_y_offset = 0
        self._scale = self._card_width / 80.
        self._sprites = Sprites(self._canvas)
        self.current_card = 0
        self._cards = []
        self._pictures = []
        self._press = None
        self._release = None
        self.timeout = None

        self._my_canvas = Sprite(
            self._sprites, 0, 0, svg_str_to_pixbuf(genblank(
                    self._width, self._height, (self._colors[0],
                                                self._colors[0]))))
        self._my_canvas.type = 'background'

        self.load_level(os.path.join(self._lessons_path, 'alphabet' + '.csv'))
        self.new_page()

    def new_page(self, cardtype='alpha'):
        ''' Load a page of cards '''
        if self.timeout is not None:
            GLib.source_remove(self.timeout)
            self.timeout = None
        self._hide_cards()
        if cardtype == 'alpha':
            self._alpha_cards()
        else:
            self._image_cards()

    def _hide_cards(self):
        if len(self._cards) > 0:
            for card in self._cards:
                card.hide()
        if len(self._pictures) > 0:
            for card in self._pictures:
                card.hide()

    def _image_cards(self):
        x = self._grid_x_offset + GUTTER
        y = self._grid_y_offset + GUTTER
        if len(self._pictures) > 0:
            for card in self._pictures:
                card.set_layer(100)
            return
        for card in self._card_data:
            self.current_card = self._card_data.index(card)
            imagefilename = self._image_data[self.current_card]
            imagepath = os.path.join(self._images_path, imagefilename)
            pixbuf = image_file_to_pixbuf(imagepath, self._card_width,
                                          self._card_height)
            self._pictures.append(Sprite(self._sprites, x, y, pixbuf))
            x += self._card_width + GUTTER * 2
            if x > self._width - (self._card_width / 2):
                x = self._grid_x_offset + GUTTER
                y += self._card_height + GUTTER * 2

    def _alpha_cards(self):
        x = self._grid_x_offset + GUTTER
        y = self._grid_y_offset + GUTTER
        if len(self._cards) > 0:
            for card in self._cards:
                card.set_layer(100)
            return
        for card in self._card_data:
            self.current_card = self._card_data.index(card)
            # Two-tone cards add some complexity.
            if type(self._color_data[self.current_card][0]) == type([]):
                stroke = self._test_for_stroke()
                top = svg_str_to_pixbuf(generate_card(
                        string=card[0],
                        colors=[self._color_data[self.current_card][0][0],
                                '#FFFFFF'],
                        scale=self._scale,
                        center=True))
                bot = svg_str_to_pixbuf(generate_card(
                        string=card[0],
                        colors=[self._color_data[self.current_card][0][1],
                                '#FFFFFF'],
                        scale=self._scale,
                        center=True))
                # Where to draw the line
                h1 = 9 / 16.
                h2 = 1.0 - h1
                bot.composite(top, 0, int(h1 * top.get_height()),
                              top.get_width(), int(h2 * top.get_height()),
                              0, 0, 1, 1, GdkPixbuf.InterpType.NEAREST, 255)
                self._cards.append(Sprite(self._sprites, x, y, top))
            else:
                stroke = self._test_for_stroke()
                self._cards.append(Sprite(self._sprites, x, y,
                                          svg_str_to_pixbuf(generate_card(
                                string='%s%s' % (
                                    card[0].upper(), card[0].lower()),
                                colors=[self._color_data[self.current_card][0],
                                        '#FFFFFF'],
                                stroke=stroke,
                                scale=self._scale, center=True))))
            x += self._card_width + GUTTER * 2
            if x > self._width - (self._card_width / 2):
                x = self._grid_x_offset + GUTTER * 2
                y += self._card_height + GUTTER * 2

    def _test_for_stroke(self):
        ''' Light colors get a surrounding stroke '''
        if self._color_data[self.current_card][0][0:4] == '#FFF':
            return True
        else:
            return False

    def new_target(self):
        self._activity.status.set_text(
            _('Click on the card that corresponds to the sound.'))
        self.target = int(uniform(0, len(self._cards)))
        # Don't repeat
        while self.target in self._deja_vu:
            self.target += 1
            self.target %= len(self._cards)
        self._deja_vu.append(self.target)
        if len(self._deja_vu) == len(self._cards):
            self._deja_vu = []
        if self.timeout is not None:
            GLib.source_remove(self.timeout)
        self.timeout = GLib.timeout_add(1000, self._play_target_sound)

    def _play_target_sound(self):
        if self._activity.mode in ['letter', 'find by letter']:
            play_audio_from_file(os.path.join(
                    self._sounds_path,
                    self._media_data[self.target][1]))
        elif self._activity.mode == 'picture':
            play_audio_from_file(os.path.join(
                    self._sounds_path,
                    self._media_data[self.target][1]))
            GLib.timeout_add(1000, play_audio_from_file, os.path.join(
                    self._sounds_path,
                    self._media_data[self.target][0]))
        else:
            play_audio_from_file(os.path.join(
                    self._sounds_path,
                    self._media_data[self.target][0]))
        self.timeout = None

    def _button_press_cb(self, win, event):
        ''' Either a card or list entry was pressed. '''
        win.grab_focus()
        x, y = list(map(int, event.get_coords()))

        spr = self._sprites.find_sprite((x, y))
        self._press = spr
        self._release = None
        return True

    def _button_release_cb(self, win, event):
        ''' Play a sound or video or jump to a card as indexed in the list. '''
        win.grab_focus()

        x, y = list(map(int, event.get_coords()))
        spr = self._sprites.find_sprite((x, y))
        if spr is None:
            return
        if spr.type == 'background':
            return
        if spr in self._cards:
            self.current_card = self._cards.index(spr)
        elif spr in self._pictures:
            self.current_card = self._pictures.index(spr)
        if self._activity.mode in ['letter', 'picture']:
            self.target = self.current_card
            self._play_target_sound()
        else:
            if self.current_card == self.target:
                self._activity.status.set_text(_('Very good!'))
                if self.timeout is not None:
                    GLib.source_remove(self.timeout)
                self.timeout = GLib.timeout_add(1000, self._correct_feedback)
            else:
                self._activity.status.set_text(_('Please try again.'))
                if self.timeout is not None:
                    GLib.source_remove(self.timeout)
                self.timeout = GLib.timeout_add(1000, self._wrong_feedback)

    def _correct_feedback(self):
        self.new_target()

    def _wrong_feedback(self):
        self._play_target_sound()

    def _keypress_cb(self, area, event):
        ''' No keyboard shortcuts at the moment. Perhaps jump to the page
        associated with the key pressed? '''
        return True

    def __draw_cb(self, canvas, cr):
        self._sprites.redraw_sprites(cr=cr)

    def _expose_cb(self, win, event):
        ''' Callback to handle window expose events '''
        self.do_expose_event(event)
        return True

    # Handle the expose-event by drawing
    def do_expose_event(self, event):

        # Create the cairo context
        cr = self._canvas.window.cairo_create()

        # Restrict Cairo to the exposed area; avoid extra work
        cr.rectangle(event.area.x, event.area.y,
                event.area.width, event.area.height)
        cr.clip()

        # Refresh sprite list
        self._sprites.redraw_sprites(cr=cr)

    def _destroy_cb(self, win, event):
        ''' Make a clean exit. '''
        Gtk.main_quit()

    def invalt(self, x, y, w, h):
        ''' Mark a region for refresh '''
        rectangle = Gdk.Rectangle()
        rectangle.x = x
        rectangle.y = y
        rectangle.width = w
        rectangle.height = h
        self._canvas.window.invalidate_rect(rectangle)

    def load_level(self, path):
        ''' Load a level (CSV) from path: letter, word, color, image,
        image sound, letter sound '''
        self._card_data = []
        self._color_data = []
        self._image_data = []
        self._media_data = []  # (image sound, letter sound)
        f = open(path)
        for line in f:
            if len(line) > 0 and line[0] not in '#\n':
                words = line.split(', ')
                self._card_data.append([words[0],
                                        words[1].replace('-', ', ')])
                if words[2].count('#') > 1:
                    self._color_data.append([words[2].split('/')])
                else:
                    self._color_data.append([words[2]])
                self._image_data.append(words[3])
                self._media_data.append((words[4], words[5]))
        f.close()

        self._clear_all()
        self._cards = []

    def _clear_all(self):
        ''' Hide everything so we can begin a new page. '''
        self._hide_cards()


def svg_str_to_pixbuf(svg_string):
    ''' Load pixbuf from SVG string. '''
    pl = GdkPixbuf.PixbufLoader.new_with_type('svg')
    pl.write(svg_string.encode("utf-8"))
    pl.close()
    return pl.get_pixbuf()


def image_file_to_pixbuf(file_path, w, h):
    ''' Load pixbuf from file '''
    return GdkPixbuf.Pixbuf.new_from_file_at_size(file_path, int(w), int(h))
